### PR TITLE
Add TopN collection

### PR DESCRIPTION
**Pull request by evgenykuzyakov**
Fix Map bug. Add Map values.

Adding missing documentation and comments.

### CODE REVIEW

File: near.ts [hunk 1]
```diff
@@ -730,6 +1001,16 @@ export namespace collections {
   export function map<K, V>(prefix: string): Map<K, V> {
     return new Map<K, V>(prefix);
   }
+
+  /**
+   * Creates or restores a persistent TopN with a given storage prefix.
+   * Always use a unique storage prefix for different collections.
+   * @param prefix A prefix to use.
+   * @
```
Comment (vgrichina) [inline near.ts]: `descending` param missing

File: near.ts [hunk 2]
```diff
@@ -16,23 +16,49 @@ export class Storage {
...
-   * NOTE: Must be very careful to avoid exploding amount of compute with this method.
-   * Make sure there is a hard limit on number of keys returned even if contract state size grows.
+   * Internal method to fetch list of keys from the given iterator up the limit.
    */
-  keys(prefix: string): string[] {
+  private _fetchIter(iterId: u32, limit: i32 = -1): string[] {
     let result: string[] = new Array<string>();
-    let iterId = storage_iter(prefix.lengthUTF8 - 1, prefix.toUTF8());
-    do {
+    while (limit-- != 0) {
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply near.ts]: that's not how `limit--` and `--limit` work

File: near.ts [hunk 3]
```diff
@@ -16,23 +16,49 @@ export class Storage {
...
+   * Returns list of keys between the given start key and the end key. Both inclusive.
+   * NOTE: Must be very careful to avoid exploding amount of compute with this method.
+   * @param start The start key used as a lower bound in lexicographical order. Inclusive.
+   * @param end The end key used as a upper bound in lexicographical order. Inclusive.
+   * @param limit The maximum number of keys to return. Default is `-1`, means no limit.
+   */
+  keyRange(start: string, end: string, limit: i32 = -1): string[] {
+    return this._fetchIter(
+      storage_range(start.lengthUTF8 - 1, start.toUTF8(), end.lengthUTF8 - 1, end.toUTF8()),
+      limit,
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply near.ts]: Personal preference, It's easier to add parameters later.

File: near.ts (hunk 1 above)
Comment (evgenykuzyakov) [reply near.ts]: done

File: near.ts [hunk 4]
```diff
@@ -704,6 +768,213 @@ export namespace collections {
...
+      this._descending = descending;
+      this._lengthKey = prefix + _KEY_LENGTH_SUFFIX;
+      this._length = -1;
+    }
+
+    /**
+     * @returns A suffix for an internal key for a given external key of type K.
+     */
+    private _keySuffix(key: K): string {
+      if (isString<K>()) {
```
[the remark this answers is not in the document]
Comment (evgenykuzyakov) [reply near.ts]: done

File: near.ts (hunk 2 above)
[the remark this answers is not in the document]
Comment (vgrichina) [reply near.ts]: @evgenykuzyakov hehe, that's why we need to change it – it's too clever )

**Pull request merged**